src/NuGet.Clients/PackageManagement.VisualStudio/Migrate/ProjectJsonToPackageRefMigrator.cs

```diff
@@ -0,0 +1,135 @@
+                var metadataElements = new List<string>();
+                var metadataValues = new List<string>();
+                if (includeFlags != LibraryIncludeFlags.All)
+                {
+                    metadataElements.Add("IncludeAssets");
+                    metadataValues.Add(LibraryIncludeFlagUtils.GetFlagString(includeFlags).Replace(',',';'));
+                }
+
+                if (privateAssetsFlag != LibraryIncludeFlagUtils.DefaultSuppressParent)
+                {
+                    metadataElements.Add("PrivateAssets");
+                    metadataValues.Add(LibraryIncludeFlagUtils.GetFlagString(privateAssetsFlag).Replace(',',';'));
+                }
+
+                await project.InstallPackageWithMetadataAsync(dependency.Name, dependency.LibraryRange.VersionRange, metadataElements, metadataValues);
+            }
+        }
+
+        private static void MigrateRuntimes(PackageSpec packageSpec, Microsoft.Build.Evaluation.Project buildProject)
+        {
+            var runtimes = packageSpec.RuntimeGraph.Runtimes;
+            var supports = packageSpec.RuntimeGraph.Supports;
+            var runtimeIdentifiers = new List<string>();
+            var runtimeSupports = new List<string>();
+            if (runtimes != null && runtimes.Count > 0)
+            {
+                runtimeIdentifiers.AddRange(runtimes.Keys);
+                
+            }
+
+            if(supports != null && supports.Count > 0)
+            {
+                runtimeSupports.AddRange(supports.Keys);
+            }
+
+            var union = string.Join(";", runtimeIdentifiers.Union(runtimeSupports));
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                buildProject.SetProperty("RuntimeIdentifiers", union);
+            });            
+        }
+
+        private static void RemoveProjectJsonReference(Microsoft.Build.Evaluation.Project buildProject, string projectJsonFilePath)
+        {
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                var projectJsonItem = buildProject.GetItems("None")
+                .Where(t => string.Equals(t.EvaluatedInclude, Path.GetFileName(projectJsonFilePath), StringComparison.OrdinalIgnoreCase))
+                .FirstOrDefault();
+
+                if (projectJsonItem != null)
+                {
+                    buildProject.RemoveItem(projectJsonItem);
+                }
+            });
+        }
+
+        private static void CreateBackup(LegacyCSProjPackageReferenceProject project,
+            string projectJsonFilePath)
+        {
+            var backupDirectory = Path.Combine(Path.GetDirectoryName(project.MSBuildProjectPath), "Backup");
+
+            Directory.CreateDirectory(backupDirectory);
```

Comment:
Given the I/O of this method, this would probably be a great method to either make sure you call from a background thread, or let this method be async and let it move itself there with `await TaskScheduler.Default;`.
Yes, at the moment your design is to block the UI thread anyway so it might not matter, except for two points:

1. It's best to avoid baking in assumptions about your caller in your individual methods. Another caller may come around someday that calls this from the UI thread in such a way as to cause a UI delay.
1. I/O on the UI thread provides an opportunity for RPC calls to get in and eventually deadlock you because it uses SynchronizationContext.Wait while it waits for the I/O to complete.